HB/common/phant-abbreviation.elpi

```diff
@@ -1,73 +1,143 @@
+typeabbrev phant-term phant.private.phant-term.
+
+namespace phant {
+
+% [add-abbreviation Name PhT C A] builds a definition "phant_Name" for the
+% term T and an abbreviation Name := phant_Name phantom-arguments as per Ph.
+% Use the API below to build a PhT as you like.
+pred add-abbreviation i:string, i:phant-term, o:constant, o:abbreviation.
+add-abbreviation N (private.phant-term AL T1) C Abbrev :- std.do! [
+  NC is "phant_" ^ N,
+  std.assert-ok! (coq.elaborate-skeleton T1 TTy T) "add-abbreviation: T illtyped",
+  log.coq.env.add-const-noimplicits NC T TTy @transparent! C,
+  private.add-abbreviation.term 0 (global (const C)) AL NParams AbbrevT,
+  @global! => log.coq.notation.add-abbreviation N NParams AbbrevT tt Abbrev,
+].
+
+% [of-gref GR ExplicitMixins PT] builds a phant-term taking all parameters,
+% the type, then inferring automatically all classing covering the mixins
+% GR depends on. ExplicitMixins is a list of mixins one wants to explicitly
+% pass (instead of being inferred)
+pred of-gref i:gref, i:list mixinname, o:phant-term.
+of-gref GRF RealMixinArgs PhBody:- !, std.do! [
+  std.assert! (gref-deps GRF MLwP) "mk-phant-term: unknown gref",
+  std.assert! (coq.env.typeof GRF FTy) "mk-phant-term: F illtyped",
+  coq.mk-eta (-1) FTy (global GRF) EtaF,
+  % toposort-mixins ML MLSorted,
+  MLwP = MLwPSorted, % Assumes we give them already sorted in dep order.
+  std.rev {list-w-params_list MLwPSorted} MLSortedRev,
+
+  std.map RealMixinArgs (m\r\ r = private.this-mixin-is-real-arg m) RMClauses,
+  std.filter MLSortedRev (m\not(std.mem! RealMixinArgs m)) MLSortedRevFiltered,
+  find-max-classes MLSortedRevFiltered CNL,
+  assert-good-coverage! MLSortedRevFiltered CNL,
+
+  RMClauses => w-params.then MLwP real-arg real-arg
+    (private.mk-phant-term.classes EtaF CNL) PhBody,
+].
+
+% A la carte API: start with a term and wrap it up
+
+% A term with no phantom arguments
+pred init i:term, o:phant-term.
+init T (private.phant-term [] T).
+
+% [real-arg N T Ph Ph1] Adds a real argument named N of type T around Ph
+pred real-arg i:name, i:term, i:(term -> phant-term), o:phant-term.
+real-arg N T F Res :- !, private.phant-fun (private.real N) T F Res.
+
+% [unify-arg Msg X1 X2 Ph Ph1] Adds an argument that will foce the unification
+% of X1 with X2 and print Msg is case of error around Ph
+pred unify-arg i:option term, i:term, i:term, i:phant-term, o:phant-term.
+unify-arg OMsg X1 X2 (private.phant-term AL F) (private.phant-term [private.idfun|AL] UF) :-
+  std.assert-ok! (coq.typecheck X1 T1) "unify-arg: X1 illtyped",
+  std.assert-ok! (coq.typecheck X2 T2) "unify-arg: X2 illtyped",
+  if (OMsg = some M) (Msg = {{lib:hb.some lp:M}}) (Msg = {{lib:hb.nomsg}}),
+  UF = {{fun unif_arbitrary : lib:hb.unify lp:T1 lp:T2 lp:X1 lp:X2 lp:Msg => lp:F}}.
+
+% [implicit-arg N T Ph Ph1] Adds an implicit argument name N of type T areound Ph
+pred implicit-arg i:name, i:term, i:(term -> phant-term), o:phant-term.
```

Comment:
I liked that the name was `.*fun.*`, because it has the signature of `fun` (except it's predicate).